_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,734 @@
+---
+layout: post
+title: "Enums: match, mutation, and moves"
+author: Felix S. Klock II
+description: "A tour of enums in Rust."
+---
+
+In this post we explore one corner of the Rust language: defining a
+class of data via `enum`, and processing instances of such data via
+`match`.
+
+Most of the post will use a single running example: modeling a game of
+chess. At the end the post changes gears, shifting to a discussion of
+how affine-typing ensures that the demonstrated features do not
+introduce unsoundness.
+
+## Tutorial material
+
+The post takes the overall form of a tutorial.
+
+If you are already familiar with defining `enums` and simple uses of
+`match`, you may want to skip ahead to the [mutating enums] or
+[avoiding unsoundness] sections. This first section is meant to
+establish everything one would need to know about enums to understand
+the discussion in latter two sections.
+
+[mutating enums]: #mutating-enums
+[avoiding unsoundness]: #avoiding-unsoundness
+
+### Defining simple enums
+
+The pieces of chess can be classified as follows:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PieceShape {
+    King, Queen, Rook, Bishop, Knight, Pawn,
+}
+```
+
+The `derive` line automatically generates support for three common
+operations: the `Copy, Clone` gives us the ability to freely copy or
+clone instances of `PieceShape`, the `Debug` provides support for
+printing each piece-rank into an output buffer, e.g. via
+`println!("piece: {:?}", piece)`, and `PartialEq, Eq` provides support
+for the `==` operator on instances of the enum.
+
+### Enum discriminant values
+
+An enum like `PieceShape` above, where each variant is just a name, is
+represented using an integer value for each variant. The definition
+style above allows the Rust compiler to assign the values itself; it
+will start from 0 and count upwards.
+
+Alternatively, one can manually assign specific values to enum
+variants, interrupting the automatic value assignment performed by the
+compiler. Thus, we could instead have written the below, and get the
+same assignment of discriminant values:
+
+```rust
+#[cfg(alternative)]
```

Comment:
I cannot tell if you are suggesting that more text be added to the post, or if you just want more context.

My method for writing this post, and my overall goal, is that someone can cut-and-paste all of the code into a file and run it.  (Indeed, my method of authorship was to run the `.md` file through a script to convert to-and-from a `.rs` file.)

So a problem arises if I want to provide alternative definitions, such as this one here, since we do not want to cause compiler errors due to duplicate definitions.